assets/src/edit-story/components/library/panes/pageLayouts/pageLayouts.js

```diff
@@ -78,19 +92,129 @@ function PageLayouts(props) {
     return { width, height, containerHeight };
   }, []);
 
+  const handleApplyPageLayout = useCallback(
+    (page) => {
+      const duplicatedPage = duplicatePage(page);
+      replaceCurrentPage({ page: duplicatedPage });
+
+      setActivePage(null);
+    },
+    [replaceCurrentPage]
+  );
+
+  const handlePageClick = useCallback(
+    (page) => {
+      if (requiresConfirmation) {
+        setIsConfirming(true);
+        setActivePage(page);
+      } else {
+        handleApplyPageLayout(page);
+        setActivePage(null);
+      }
+    },
+    [requiresConfirmation, handleApplyPageLayout]
+  );
+
+  const handleKeyboardPageClick = useCallback(
+    ({ key }, page) => {
+      if (key === 'Enter') {
+        if (isPageLayoutsFocused) {
+          handlePageClick(page);
+        }
+      }
+    },
+    [isPageLayoutsFocused, handlePageClick]
+  );
+
+  const handleCloseDialog = useCallback(() => {
+    setActivePage(null);
+    setIsConfirming(false);
+  }, [setIsConfirming]);
+
+  const handleConfirmDialog = useCallback(() => {
+    handleApplyPageLayout(activePage);
+    setIsConfirming(false);
+  }, [activePage, handleApplyPageLayout]);
+
+  const handlePageFocus = useCallback(
+    (pageId) => {
+      if (activePageId !== pageId) {
+        setActivePageId(pageId);
+      }
+    },
+    [activePageId]
+  );
+
   const rowVirtualizer = useVirtual({
-    size: Math.ceil(pages.length / 2),
+    size: Math.ceil((pages || []).length / 2),
     parentRef,
     estimateSize: useCallback(
       () => pageSize.containerHeight + PAGE_LAYOUT_ROW_GAP,
       [pageSize.containerHeight]
     ),
-    overscan: 2,
+    overscan: 4,
   });
 
-  const requiresConfirmation = useMemo(
-    () => currentPage && !isDefaultPage(currentPage),
-    [currentPage]
+  const columnVirtualizer = useVirtual({
+    horizontal: true,
+    size: 2,
+    parentRef,
+    estimateSize: useCallback(() => pageSize.width + PAGE_LAYOUT_ROW_GAP, [
+      pageSize.width,
+    ]),
+    overscan: 0,
+  });
+
+  // this is how we make sure the proper ref is focused when more rows are loaded
+  // otherwise focus gets thrown off
+  // checking for a difference in currentAvailableRows and assigning the matching ref to any change
+  // is just so that the effect hook will run when the virtualized list updates and reattach focus to the new instance
+  // of an already selected layout.
+  const currentAvailableRows = useMemo(() => rowVirtualizer.virtualItems, [
+    rowVirtualizer,
+  ]);
+  const currentAvailableRowsRef = useRef();
+
+  useEffect(() => {
+    if (currentAvailableRows !== currentAvailableRowsRef?.current) {
+      currentAvailableRowsRef.current = currentAvailableRows;
+    }
+
+    if (activePageId && isPageLayoutsFocused) {
+      pageRefs.current?.[activePageId]?.focus();
+    }
+  }, [activePageId, currentAvailableRows, isPageLayoutsFocused]);
```

Comment:
this is where the keyboard focus magic happens.

Reply:
This is nuts to track focus like this, but seems to work well so nice work! Could you add some more comments around this and how focus is tracked through `pageRefs` and `isPageLayoutsFocused`? I would not understand why this was all happening if I didn't know the context of testing this ticket.